### PR TITLE
H-33: Use RDS for Temporal

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Currently, we use a Docker container to run Postgres. This changes the Postgres instance to live in RDS instead

## 🔍 What does this change?

- Setup Postgres rules for `temporal` and `temporal_visibility`.
- Associate the Temporal service the RDS instance and remove old docker container
- For the purpose of easier and more secure testing I limited the Temporal service to the CIDR block

### :ship: Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### :scroll: Does this require a change to the docs?

The changes in this PR:

- [x] is internal and does not require a docs change

### :spider_web: Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] does not affect the execution graph

## 🐾 Next steps

Probably
- Setup authentication
- Setup visibility service
- Address TODOs in Temporal TF code

### CODE REVIEW

File: infra/terraform/hash/main.tf [hunk 1]
```diff
@@ -92,10 +92,20 @@ module "temporal" {
   vpc                   = module.networking.vpc
   env                   = local.env
   region                = local.region
-  cpu                   = 512
-  memory                = 1024
+  cpu                   = 256
+  memory                = 512
```
Comment (TimDiekmann) [inline infra/terraform/hash/main.tf:96]: We can go higher as we need to. For the current purpose it's fine to stick with the lowest spec.

File: infra/terraform/hash/main.tf [hunk 2]
```diff
@@ -92,10 +92,20 @@ module "temporal" {
...
   # TODO: provide by the HASH variables.tf
   temporal_version      = "1.21.0.0"
+
+  postgres_host = module.postgres.pg_host
+  postgres_port = module.postgres.pg_port
+  postgres_db = "temporal"
+  postgres_visibility_db = "temporal_visibility"
+  postgres_user = "temporal"
```
Comment (TimDiekmann) [inline infra/terraform/hash/main.tf:104]: We may want to parameterize these as well

File: infra/terraform/hash/postgres_roles/postgres_roles.tf [hunk 3]
```diff
@@ -135,4 +135,52 @@ resource "postgresql_database" "graph" {
...
+}
+
+resource "postgresql_default_privileges" "temporal_readwrite_tables" {
+  owner       = var.pg_superuser_username
+  role        = postgresql_role.temporal_user.name
+  database    = postgresql_database.temporal.name
+  schema      = "public"
+
+  object_type = "table"
+  privileges  = ["SELECT", "INSERT", "UPDATE", "DELETE"]
```
Comment (TimDiekmann) [inline infra/terraform/hash/postgres_roles/postgres_roles.tf:164]: We probably don't need all CRUD operations (same for the visibility)

File: infra/terraform/modules/temporal/main.tf [hunk 4]
```diff
@@ -174,8 +183,9 @@ resource "aws_security_group" "app_sg" {
     from_port   = 7233
     to_port     = 7233
     protocol    = "tcp"
-    description = "Allow connections to Temporal from anywhere (rely on authentication to restrict access)"
-    # TODO: Consider changing this to `var.vpc.cidr_block`
-    cidr_blocks = ["0.0.0.0/0"]
+    description = "Allow connections to Temporal within the VPC"
+    # TODO: Consider changing this to `"0.0.0.0/0"` and setup authentication
+    # description = "Allow connections to Temporal from anywhere (rely on authentication to restrict access)"
+    cidr_blocks = [var.vpc.cidr_block]
```
Comment (TimDiekmann) [inline infra/terraform/modules/temporal/main.tf:189]: Changed for more secure testing

File: infra/terraform/modules/temporal/task_definitions.tf [hunk 5]
```diff
@@ -154,28 +110,39 @@ locals {
...
-  secret_raw_vars = [
-    # TODO: parameterize module
-    { name = "POSTGRES_USER", value = "postgres" },
-    { name = "POSTGRES_PWD", value = "postgres" },
-    { name = "POSTGRES_SEEDS", value = "localhost" },
+  temporal_migration_env_vars = [
+#    { name = "POSTGRES_USER", value = var.postgres_user },
+    { name = "POSTGRES_USER", value = var.postgres_superuser },
+#    { name = "SKIP_DB_CREATE", value = "true" },
+#    { name = "SKIP_VISIBILITY_DB_CREATE", value = "false" },
```
Comment (TimDiekmann) [inline infra/terraform/modules/temporal/task_definitions.tf:126]:
```suggestion
    { name = "POSTGRES_USER", value = var.postgres_superuser },
```

File: infra/terraform/modules/temporal/task_definitions.tf [hunk 6]
```diff
@@ -154,28 +110,39 @@ locals {
...
-  shared_secrets = [for env_name, ssm_param in aws_ssm_parameter.secret_env_vars :
-  { name = env_name, valueFrom = ssm_param.arn }]
-}
+  temporal_env_vars = [
+    { name = "POSTGRES_USER", value = var.postgres_user },
+  ]
 
+  temporal_migration_secrets = [
+    { name = "POSTGRES_PWD", value = var.postgres_superuser_password },
+#    { name = "POSTGRES_PWD", value = var.postgres_password },
```
Comment (TimDiekmann) [inline infra/terraform/modules/temporal/task_definitions.tf:135]:
```suggestion
```

**Review by CiaranMn — APPROVED** [review]

**Pull request merged**